File: microservice-architecture/idempotency-pattern/ts/test/test_idempotency.ts
import { IdempotencyOrderService } from "../src/idempotency";

function assertEqual<T>(expected: T, actual: T, message: string): void {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`);
  }
}

const service = new IdempotencyOrderService();

const first = service.createOrder("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 1);
assertEqual("CREATED", first.status, "first request should create");
assertEqual(false, first.replayed, "first request should not be replayed");

const second = service.createOrder("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 1);
assertEqual("CREATED", second.status, "duplicate request should replay");
assertEqual(true, second.replayed, "duplicate request should be replayed");

const conflict = service.createOrder("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 2);
assertEqual("CONFLICT", conflict.status, "different payload should conflict");
assertEqual(false, conflict.replayed, "conflict should not be replayed");

console.log("idempotency-pattern(ts) tests passed");

File: microservice-architecture/idempotency-pattern/python/test/test_idempotency.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.idempotency import IdempotencyOrderService  # noqa: E402


class IdempotencyPatternTest(unittest.TestCase):
    def test_idempotency_pattern(self) -> None:
        service = IdempotencyOrderService()

        first = service.create_order("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 1)
        self.assertEqual("CREATED", first.status)
        self.assertFalse(first.replayed)

        second = service.create_order("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 1)
        self.assertEqual("CREATED", second.status)
        self.assertTrue(second.replayed)

        conflict = service.create_order("IDEMP-ORDER-1001", "ORD-1001", "SKU-BOOK", 2)
        self.assertEqual("CONFLICT", conflict.status)
        self.assertFalse(conflict.replayed)


if __name__ == "__main__":
    unittest.main()
